### ISSUE TITLE
Folding vs Lowering: improving the compiler flow

### ISSUE THREAD

**Issue by tlepley-cadence** [pytorch/glow#2615]
The problematic is the following:
- input models (ONNX for instance) can express the exact same functionality with various forms. Examples are LeakyRelu and BatchNorm that may be expressed as a sub-graph rather than a single node in the input model (We've got this situation multiple times)
- Backend may or may not support a high-level node. This should be handled by the lowering phase that is under their control.

@artemrakhov  proposes in #2613 (for LeakyRelu) to have have a high level operator lowered directly in the high level IR, before lowering, and then to request backends to fold high-level nodes when they support high level function natively (then to do highering after lowering). 

I don't think this approach should be followed. My proposal is actually to do the opposite: have the high-level IR as high-level as possible and let the lowering pass perform the IR lowering, as expected, driven by backends. 

There are IMO many advantages:
- The high level IR more canonical, what simplifies many optimizations. Another example is for instance BatchNorm that we have often seen unfolded in various ways in converted ONNX models.
- it makes the compiler flow more consistent, going from high level (high abstraction) to low (level/low abstraction), stage after stage.
- it simplifies backends and make the high-level IR more independent from the CPU backend specificities. Many backends will support high-level nodes natively, on the contrary to the CPU backend.
- it make the IR more consistent. Today, for instance, we have Relu in the high IR, so it looks consistent to have LeakyRelu too.

A solution that is IMO clean consists of splitting the `glow::optimize()` in 2 parts:
1/ IR 'legalizer' *called only once* at the beginning of the compiler flow. The purpose is to make the high level IR more independent from the input models variants. 
2/ Generic optims (sink, merge etc...) *called before and after lowering* like today.

Then, we would have as compiler flow:

1- import model
2- legalize the IR (fold high-level nodes as much as possible)
3- generic optims (no folding)
4- lowering
5- generic optims (no folding)

**Comment by jfix71** [comment]
I like the idea in general. We have this same situation in our en2gr example, where the LSTMs are already lowered in the proto/when imported.

It does make things much cleaner to "raise" (called legalize here) the LSTM back up to a high level LSTM node, e.g. if we wanted to write high-level optimizations for LSTMs, or if a backend wants to understand/consume it as an LSTM. And I do think it fits more with Glow's philosophy -- "raise" as much as possible at the very beginning, then lower only as necessary as we move down the compilation pipeline. If we wait to raise after optimizations, the graph may be optimized into a form that makes it more complicated to pattern match.

I also think it makes sense to make this logic independent of a backend. One consideration here is to make this raising opt-in by the backend (e.g. `Backend::shouldRaise()`?), especially if we have no graph optimizations that depend on the high-level node. This is because e.g. if the target backend does want an LSTM lowered, and we don't have any optimizations that depend on it being raised back to an LSTM, then we're essentially wasting work to raise it and then lower it again with no benefit.

**Comment by tlepley-cadence** [comment]
In #2613, we agreed to have a folding pass early in the flow. The main purpose is the 'higher' the IR in case the input models is malformed. Currently, the call to this folding pass is called in `Loader`.

**Comment by jfix71** [comment]
@tlepley-cadence Closing this as I think #2613 fully addresses it.
